File: core1/us/trader/workflow/organization/workflow_t.cpp
//===-                           P L E B B L E
//===-                         https://plebble.us
//===-
//===-              Copyright (C) 2017-2022 root1m3@plebble.us
//===-
//===-                      GNU GENERAL PUBLIC LICENSE
//===-                       Version 3, 29 June 2007
//===-
//===-    This program is free software: you can redistribute it and/or modify
//===-    it under the terms of the GPLv3 License as published by the Free
//===-    Software Foundation.
//===-
//===-    This program is distributed in the hope that it will be useful,
//===-    but WITHOUT ANY WARRANTY; without even the implied warranty of
//===-    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
//===-
//===-    You should have received a copy of the General Public License
//===-    along with this program, see LICENCE file.
//===-    see https://www.gnu.org/licenses
//===-
//===----------------------------------------------------------------------------
//===-
#include "workflow_t.h"
#include "docs.h"

#define loglevel "trader/workflow/organization"
#define logclass "workflow_t"
#include <us/gov/logs.inc>

using namespace us::trader::workflow::organization;
using c = us::trader::workflow::organization::workflow_t;

c::workflow_t() {
}

c::~workflow_t() {
}

c::bitem* c::enable_appointment(bool b, ch_t& ch) {
    return b ? add<appointment_t, 1>(ch) : remove<appointment_t>(ch);
}


c::bitem* c::enable_reference(bool b, ch_t& ch) {
    return b ? add<reference_t, 2>(ch) : remove<reference_t>(ch);
}

namespace {

    using namespace std;
    using namespace us;

    template<typename a, c::item_factory_id_t item_factory_id>
    struct my_item_factory_t: c::item_factory_t {
        pair<ko, value_type*> create() const override {
            return make_pair(ok, new item_t<a, item_factory_id>());
        }
    };

}

void c::register_factories(item_factories_t& item_factories) const {
    item_factories.register_factory(appointment_factory_id, new my_item_factory_t<appointment_t, appointment_factory_id>());
    item_factories.register_factory(reference_factory_id, new my_item_factory_t<reference_t, reference_factory_id>());
}

File: core0/us/android/app/src/main/java/us/cash/workflow_t.java
//===-                           P L E B B L E
//===-                         https://plebble.us
//===-
//===-              Copyright (C) 2017-2022 root1m3@plebble.us
//===-
//===-                      GNU GENERAL PUBLIC LICENSE
//===-                       Version 3, 29 June 2007
//===-
//===-    This program is free software: you can redistribute it and/or modify
//===-    it under the terms of the GPLv3 License as published by the Free
//===-    Software Foundation.
//===-
//===-    This program is distributed in the hope that it will be useful,
//===-    but WITHOUT ANY WARRANTY; without even the implied warranty of
//===-    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
//===-
//===-    You should have received a copy of the General Public License
//===-    along with this program, see LICENCE file.
//===-    see https://www.gnu.org/licenses
//===-
//===----------------------------------------------------------------------------
//===-
package us.cash;
import java.util.ArrayList;                                                                    // ArrayList
import android.content.Context;                                                                // Context
import us.wallet.trader.data_t;                                                                // data_t
import java.util.HashMap;                                                                      // HashMap
import static us.gov.crypto.ripemd160.hash_t;                                                  // hash_t
import static us.stdint.*;                                                                     // *
import static us.tuple.*;                                                                      // *
import us.ko;                                                                                  // ko
import java.util.Map;                                                                          // Map
import static us.ko.ok;                                                                        // ok
import us.pair;                                                                                // pair
import android.view.View;                                                                      // View

public class workflow_t extends HashMap<String, pair<workflow_item_t, workflow_item_t>> { //local, remote

    static void log(final String s) { //--strip
        System.out.println("workflow: " + s); //--strip
    } //--strip

    public workflow_t(data_t data) {
        log("workflow_t constructor"); //--strip
        for (Map.Entry<String, String> i: data.entrySet()) {
            if (i.getKey().startsWith("local__wf_")) {
                add(0, i.getKey().substring(10), i.getValue());
                continue;
            }
            if (i.getKey().startsWith("remote__wf_")) {
                add(1, i.getKey().substring(11), i.getValue());
                continue;
            }
        }
    }

    void add(int section, String prop, String value) {
        log("add " + section + " prop:" + prop + " val: " + value); //--strip
        String[] docprop = prop.split("_", 2);
        log("docprop " + docprop.length + " doc:" + docprop[0] + " val: " + value); //--strip
        pair<workflow_item_t, workflow_item_t> i = get(docprop[0]);
        if (i == null) {
            i = new pair<workflow_item_t, workflow_item_t>(new workflow_item_t(docprop[0]), new workflow_item_t(docprop[0]));
            put(docprop[0], i);
        }
        String wprop = null;
        if (docprop.length == 2) {
            wprop = docprop[1];
        }
        if (section == 0) {
            i.first.add(wprop, value);
        }
        else {
            i.second.add(wprop, value);
        }

    }

/*
    public void refresh(trader.data_t data) {
        for (Map.Entry<String, workflow_item> set : entrySet()) {
            set.getValue().refresh(data);
        }
    }

    public void add_redirection() {
        put("redir", new workflow_item_redirect(ctx, v, tr));
    }
*/
//    trader tr;

}

